File: datastructures/linkedlist/python/algorithms/classic/palindrome_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def is_palindrome(head):
    if not head:
        return True

    second_half_head = find_second_half_head(head)
    reversed_second_half_head = reverse(second_half_head.next)

    p1 = head
    p2 = reversed_second_half_head

    while p1 and p2:
        if p1.val != p2.val:
            return False

        p1 = p1.next
        p2 = p2.next

    second_half_head.next = reverse(reversed_second_half_head)

    return True


def find_second_half_head(head):
    slow = head
    fast = head

    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next

    return slow


def reverse(head):
    prev = None

    while head:
        next = head.next
        head.next = prev
        prev = head
        head = next

    return prev


def print_list(head):
    current = head

    while current:
        print(current.val)
        current = current.next


if __name__ == "__main__":
    node_1 = ListNode(1)
    node_2 = ListNode(2)
    node_3 = ListNode(3)
    node_4 = ListNode(4)
    node_5 = ListNode(3)
    node_6 = ListNode(2)
    node_7 = ListNode(1)

    node_1.next = node_2
    node_2.next = node_3
    node_3.next = node_4
    node_4.next = node_5
    node_5.next = node_6
    node_6.next = node_7

    print(is_palindrome(node_1))
    print_list(node_1)
    print(" ")

File: datastructures/linkedlist/java/algorithms/classic/PalindromeLinkedList.java
public class PalindromeLinkedList {

    public static boolean isPalindrome(ListNode head) {
        if (head == null) {
            return true;
        }

        ListNode secondHalfHead = findSecondHalf(head);
        ListNode reversedSecondHalfHead = reverse(secondHalfHead.next);

        ListNode p1 = head;
        ListNode p2 = reversedSecondHalfHead;

        while (p1 != null && p2 != null) {
            if (p1.val != p2.val) {
                return false;
            }

            p1 = p1.next;
            p2 = p2.next;
        }

        // Restore the list
        secondHalfHead.next = reverse(reversedSecondHalfHead);

        return true;
    }

    private static ListNode findSecondHalf(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast.next != null && fast.next.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return slow;
    }

    private static ListNode reverse(ListNode head) {
        ListNode prev = null;

        while (head != null) {
            ListNode next = head.next;
            head.next = prev;
            prev = head;
            head = next;
        }

        return prev;
    }

    private static void print(ListNode head) {
        ListNode current = head;

        while (current != null) {
            System.out.print(current.val + " ");
            current = current.next;
        }
    }

    public static void main(String[] args) {
        ListNode node1 = new ListNode(1);
        ListNode node2 = new ListNode(2);
        ListNode node3 = new ListNode(3);
        ListNode node4 = new ListNode(4);
        ListNode node5 = new ListNode(3);
        ListNode node6 = new ListNode(2);
        ListNode node7 = new ListNode(1);

        node1.next = node2;
        node2.next = node3;
        node3.next = node4;
        node4.next = node5;
        node5.next = node6;
        node6.next = node7;

        print(node1);

        System.out.println(isPalindrome(node1));

        print(node1);
    }

}
